crates/jstz_core/src/kv/outbox.rs

```diff
@@ -0,0 +1,461 @@
+
+        assert_eq!(2, num_flushed);
+        assert_eq!(0, outbox_queue.queue_len());
+
+        let level = host.run_level(|_| {});
+        let outbox = host.outbox_at(level);
+
+        for (i, message) in outbox.iter().enumerate() {
+            let (_, message) =
+                OutboxMessageFull::<OutboxMessage>::nom_read(message).unwrap();
+            assert_eq!(message, make_withdrawal(&accounts[i]).into());
+        }
+    }
+
+    #[test]
+    fn flush_rollup_queue_first_then_snapshot_queue() {
+        let mut host = MockHost::default();
+        let mut outbox_queue = OutboxQueue::default();
+        outbox_queue.meta.queue_len = 4;
+
+        let accounts = [
+            PublicKeyHash::digest(b"account1").unwrap(),
+            PublicKeyHash::digest(b"account2").unwrap(),
+            PublicKeyHash::digest(b"account3").unwrap(),
+            PublicKeyHash::digest(b"account4").unwrap(),
+        ];
+        for i in 0..2 {
+            outbox_queue
+                .rollup_outbox_queue
+                .queue_message(&mut host, make_withdrawal(&accounts[i]))
+                .unwrap();
+        }
+
+        let outbox_queue_snapshot = OutboxQueueSnapshot(
+            accounts[2..]
+                .iter()
+                .map(|acc| make_withdrawal(acc))
+                .collect(),
+        );
+
+        let num_flushed = outbox_queue.flush(&mut host, outbox_queue_snapshot);
+
+        assert_eq!(0, outbox_queue.queue_len());
+        assert_eq!(4, num_flushed);
+
+        let level = host.run_level(|_| {});
+        let outbox = host.outbox_at(level);
+
+        assert_eq!(4, outbox.len());
+
+        for (i, message) in outbox.iter().enumerate() {
+            let (_, message) =
+                OutboxMessageFull::<OutboxMessage>::nom_read(message).unwrap();
+            assert_eq!(message, make_withdrawal(&accounts[i]).into());
+        }
+    }
+
+    #[test]
+    fn flush_enqueues_remaining_messages_to_rollup_queue() {
+        let mut host = MockHost::default();
+        let mut messages: Vec<OutboxMessage> = vec![];
+        for i in 0..120 {
+            let account =
+                PublicKeyHash::digest(format!("account{}", i).as_bytes()).unwrap();
+            messages.push(make_withdrawal(&account))
+        }
+        let mut messages = messages.into_iter();
+
+        let max = u16::MAX as u32;
+        let mut outbox_queue = OutboxQueue {
+            meta: OutboxQueueMeta {
+                queue_len: 120,
+                max,
+            },
+            rollup_outbox_queue: RollupOutboxQueue::new(&ROLLUP_OUTBOX_QUEUE_ROOT, max)
+                .unwrap(),
+        };
+
+        for message in messages.by_ref().take(60) {
+            OUTBOX_QUEUE.queue_message(&mut host, message).unwrap();
+        }
+
+        let outbox_queue_snapshot = OutboxQueueSnapshot(messages.take(60).collect());
+
+        outbox_queue.flush(&mut host, outbox_queue_snapshot);
+
+        assert_eq!(20, outbox_queue.queue_len());
```

Comment:
Incorrect calculation: 120 - 60 (flushed) - 40 (written) = 20
Should be: 60 - 60 (flushed) + 20 (queued) = 20